### ISSUE TITLE
OpenMM minimizer failing

### ISSUE THREAD

**Issue by stefdoerr** [pandegroup/openmm#1857]
I promised an example where the Acemd minimizer succeeds yet the OpenMM minimizer won't.

After minimization, the system won't simulate in neither software (probably due to clashing atoms). In Acemd because of barostat issues and in OpenMM because it's sending my atoms to space.

But still, talking only about the minimizer here is an example. The openMM python stuff I stole from CHARMM-GUI which produces these python scripts.

You can do `conda install acemd -c acellera` to get acemd.

Then try with `python simulate_openmm.py`

Initial system energy
2.933086022969151e+22 kJ/mol
3.3277619823620913e+22 kJ/mol

And then try by calling `acemd`
You will see that Acemd is able to bring down the energy by a lot if you reload the output.xtc and evalute the energy in openMM.

[example.zip](https://github.com/pandegroup/openmm/files/1148834/example.zip)

I am also tagging Raimondas here as I am going today on holidays and he will probably be working more with openMM in the future @raimis .

**Comment by peastman** [comment]
Which platform are you using?  The CUDA and OpenCL platforms accumulate forces in 64 bit fixed point.  That means if any component of the force on any atom is larger than about +/- 2 billion kJ/mol/nm, it overflows and you get a nan.  That's not an issue during normal simulations (real forces are many orders of magnitude smaller than that), but given that initial energy, your starting structure probably does have forces that high.  Try minimizing with the CPU platform, which doesn't have that issue.

Still, I wouldn't expect the minimizer to be able to do much with a starting structure that bad.  It sounds like the Acemd one isn't able to either, since you can't run a simulation from the structure it produces.

**Comment by stefdoerr** [comment]
Yes that was with CUDA but we have similar issues with simple protein systems which actually simulate in Acemd afterwards. Raimondas could maybe provide the 3ptb example that we have on our issue tracker when he finds some time.

On July 14, 2017 6:52:59 PM GMT+02:00, peastman <notifications@github.com> wrote:
>Which platform are you using?  The CUDA and OpenCL platforms accumulate
>forces in 64 bit fixed point.  That means if any component of the force
>on any atom is larger than about +/- 2 billion kJ/mol/nm, it overflows
>and you get a nan.  That's not an issue during normal simulations (real
>forces are many orders of magnitude smaller than that), but given that
>initial energy, your starting structure probably does have forces that
>high.  Try minimizing with the CPU platform, which doesn't have that
>issue.
>
>Still, I wouldn't expect the minimizer to be able to do much with a
>starting structure that bad.  It sounds like the Acemd one isn't able
>to either, since you can't run a simulation from the structure it
>produces.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/pandegroup/openmm/issues/1857#issuecomment-315409734

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by peastman** [comment]
And what happens if you use CPU?

**Comment by stefdoerr** [comment]
No clue. I only tried with GPU. Im at the airport now. You can try though if you want with the files I attached.

On July 14, 2017 7:23:51 PM GMT+02:00, peastman <notifications@github.com> wrote:
>And what happens if you use CPU?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/pandegroup/openmm/issues/1857#issuecomment-315416817

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by peastman** [comment]
Give it a try when you have a chance and see if it works.

**Comment by stefdoerr** [comment]
Indeed with CPU it works:
```
Initial system energy
2.9391711278153706e+22 kJ/mol
-284777.0506019093 kJ/mol  # after minimization
```
So can I minimize with CPU and then change platform to CUDA for the rest of the simulation?

**Comment by giadefa** [comment]
We would need a minimazation on GPUs most of the times. Any reason why it
breaks there?

**Comment by stefdoerr** [comment]
floating point single precision as far as I understand is not enough to
hold such huge values so it cannot minimize correctly when starting with
such energies.

**Comment by giadefa** [comment]
GPUs are not restricted to single though, mixed could be a solution or a
simpler first order scheme.

**Comment by peastman** [comment]
It's accumulating forces in 64 bit fixed point format.  That's much faster than double precision, and much more accurate than single precision.  But it does set a limit on how large any force component can be before it overflows.

> So can I minimize with CPU and then change platform to CUDA for the rest of the simulation?

Yes.  After you minimize, just call `getState(getPositions=True)` on the Context.  Then create a new Context with the CUDA platform and call `setState()` on it to give it the minimized positions.

> We would need a minimazation on GPUs most of the times.

Why?  Using the CPU platform will be slower, but minimization isn't usually a limiting factor.  If you're worried about speed, you could do only a small amount of minimization on the CPU, just to bring the largest forces down to a reasonable range.  For example, you could specify `tolerance=1e6`.  Then finish up the minimization on the GPU.

**Comment by stefdoerr** [comment]
Awesome! Many thanks for the tip on how to switch context.

**Comment by giadefa** [comment]
HI Peter,
is there a way to renormalize the forces beyond the maximum so to make sure
that it is robust? Even if optional?

It is a bit of a problem the fact that one cannot rely on it on GPUs.
Although minimisation is usually less than a thousand steps we plan to do
it on thousands of systems.

Best,
g

**Comment by stefdoerr** [comment]
The huge energy in this case Gianni I think is due to a ring penetration of
a lipid.
Once that's solved, the energy will probably go down to more normal values.

**Comment by jchodera** [comment]
I wonder if performing the first few steps of minimization with `double` precision on the CUDA platform would provide more robustness without slowing things down too much. You'd have to create a separate `Context` for `double`, but it may still be much faster than using the `CPU` platform.

**Comment by peastman** [comment]
Switching to double wouldn't help.  The problem comes from how forces are accumulated, not how individual forces are calculated.  And accumulating in double precision isn't an option, because there aren't atomic operations for it.

But here's something you might try.  The OpenCL platform has the same mechanism of accumulating forces in fixed point, but it also has an alternate mechanism that accumulates them with lots of separate floating point buffers.  That's slower and uses a lot more memory. but it's needed on low end GPUs that don't support 64 bit atomics.  It automatically selects which mechanism to use based on your GPU's feature set, but you could easily modify it to change that.  Just before this line:

https://github.com/pandegroup/openmm/blob/master/platforms/opencl/src/OpenCLContext.cpp#L253

insert the following:

```C++
supports64BitGlobalAtomics = false;
```

See if that works for you.

**Comment by jchodera** [comment]
It looks like [atomic add operations are now supported for doubles in SM_6X (Pascal) GPUs as of CUDA 8](https://stackoverflow.com/questions/12626096/why-has-atomicadd-not-been-implemented-for-doubles).

**Comment by stefdoerr** [comment]
Update. Even resolving the ring penetrations doesn't bring the energy down to manageable levels since there are lots of atom clashes in the membrane. Works fine with CPU. So I will try now the context switching from CPU minimization to GPU simulation.

Even though as Gianni said ideally we would minimize ~10 steps on CPU and the rest of the minimization+simulation on the GPU or some other fix for handling large energies.

**Comment by stefdoerr** [comment]
I managed to make it work with CPU minimization + switching to GPU simulations.

But then I thought I would try doing 10 steps minimization on CPU and then switch context to GPU minimization + simulation, but I am seeing some curious behaviour with the minimizer

```C
OpenMM::LocalEnergyMinimizer::minimize( *(sim->minimization_context), 100, 10 );
sim->context->setPositions(sim->minimization_context->getState(OpenMM::State::Positions).getPositions());
OpenMM::LocalEnergyMinimizer::minimize( *(sim->context), 100, 500000 );
```

The `minimization_context` runs on CPU and the `context` runs on GPU.
If I do this, once it switches context it doesn't minimize anymore (maybe the energies are still too large?)
But If I put like 50 maxiterations it keeps running for way too long (and works fine afterwards).
So I cannot find exactly the middle ground that takes little time and minimizes correctly.

Would be helpful if the minimizer had a verbosity option to show the energy at each step.

**Comment by peastman** [comment]
Instead of limiting it by max iterations, try limiting it by tolerance.  If you set it to `1e6`, that will do enough steps to bring it down into a range where GPU should have no problems, but also won't work at it too hard.

Alternatively, you could just use an intermediate number of steps.  You already found that 10 is too few and 50 is too many.  So what about 20?  But specifying the tolerance is probably a better way, since it will figure this out for you automatically.

**Comment by stefdoerr** [comment]
Sorry but I'll have to necro this issue. I have a docked ligand which in one spot is a bit close to the protein. The minimizer doesn't move pretty much anything and is unable to bring the energy down even on CPU. It also seems to be extremely consistent no matter number of iterations or any other options at minimizing to the exact same structure.

I attach the files. [files.zip](https://github.com/pandegroup/openmm/files/1462022/files.zip)

Here is the code to replicate the issue. You can remove the HTMD parts, I only used it for reading my coordinates from the PDB.
```python
import simtk.unit as unit
import simtk.openmm as mm
from simtk.openmm import app
import parmed
from htmd.molecule.molecule import Molecule
import numpy as np


def getPlatform(plat, device):
    platform = mm.Platform.getPlatformByName(plat)
    prop = None
    if plat == 'CUDA':
        prop = {'CudaPrecision': 'single', 'CudaDeviceIndex': str(device)}
    return platform, prop


prmtop = './structure.prmtop'
mol = Molecule('./structure.pdb')
coords = mol.coords.squeeze()

coordsOMM = unit.Quantity(coords.copy(), unit=unit.angstrom)
parm = parmed.amber.AmberParm(prmtop)
system = parm.createSystem(nonbondedMethod=app.CutoffNonPeriodic)

platform, prop = getPlatform('CPU', 0)
integrator = mm.LangevinIntegrator(300 * unit.kelvin, 1.0 / unit.picoseconds, 2.0 * unit.femtoseconds)
simulation = app.Simulation(parm.topology, system, integrator, platform, prop)
simulation.context.setPositions(coordsOMM)

state = simulation.context.getState(getEnergy=True, getForces=True)
print('Starting pot energy:', state.getPotentialEnergy().in_units_of(unit.kilocalories_per_mole))
simulation.minimizeEnergy(tolerance=0, maxIterations=10000)
state = simulation.context.getState(getPositions=True, getEnergy=True, getForces=True)
print('Ending pot energy:', state.getPotentialEnergy().in_units_of(unit.kilocalories_per_mole))
newcoor = state.getPositions(asNumpy=True).value_in_unit(unit.angstrom)
print(np.max(np.abs(newcoor - coords)))
```


```
Starting pot energy: 202405.93606898971 kcal/mol
Ending pot energy: 202350.01280613983 kcal/mol
0.0584910831375
```

**Comment by stefdoerr** [comment]
My point being that there is still a quite strong clash between the protein and the ligand after minimization. Here is the energy breakdown.
```
HarmonicBondForce 611.5656890289684
HarmonicAngleForce 854.6532854514649
PeriodicTorsionForce 3473.1117921858013
NonbondedForce 186172.19753015545
CustomGBForce -4382.068978101575
CMMotionRemover 0.0
```

**Comment by peastman** [comment]
It looks like you have a really bad clash the minimizer isn't able to resolve.  I added the following lines at the end of your script:

```python
for i, f in enumerate(state.getForces()):
  if unit.norm(f) > 1e6*unit.kilojoules_per_mole/unit.nanometer:
    print(i+1, f)
```

It prints the following:

```
786 (12513539.19991997, -53160522.255137384, 1746220.629537553) kJ/(nm mol)
4813 (-11665205.082050947, -8186013.236372756, -13719900.60820041) kJ/(nm mol)
4819 (6932218.352257314, -1181790.3621584717, 5192165.5394199565) kJ/(nm mol)
4820 (5378622.865030534, 9481147.34778779, 8736793.783853194) kJ/(nm mol)
4845 (-13454407.032574195, 11118994.076644477, 793385.9973595268) kJ/(nm mol)
4866 (4527450.853239046, 43100670.51801402, -3683295.9232167997) kJ/(nm mol)
4867 (-3781863.4258160894, -896869.9685525877, 1052793.3702675407) kJ/(nm mol)
```

Looking up those atom indices in the PDB file, I see that several of them are part of the MOL residue, which is a good clue that the ligand has probably been positioned badly.  Pulling it up in VMD and highlighting those atoms, I see that sure enough, they're sitting right on top of part of the protein.

**Comment by jchodera** [comment]
Could simply shrinking the LJ radii help?

**Comment by stefdoerr** [comment]
Unzip this [input.zip](https://github.com/pandegroup/openmm/files/1462948/input.zip) into the same folder and do `acemd input` 
(if you don't have acemd you can do `conda install acemd`)
Minimizes fine. 

I mean I'm really sorry if I keeping on making this argument. I'm not trying to compare software, it's just that I'd really like to stop using ACEMD for stuff I can do in python with openMM. But for my applications this kills the usability of the software if it doesn't work in 1/10th of my systems.

![image](https://user-images.githubusercontent.com/7935362/32682087-fe64f8d6-c672-11e7-9f66-519160a22c13.png)

**Comment by stefdoerr** [comment]
The original distance of the clashing atom pair is 0.91 A, but also in the correct direction (i.e. one it can escape to). This system minimizes with other simulation software as well (desmond I think).

**Comment by jchodera** [comment]
For acemd and Desmond, are you using the exact same forcefield? And it's just that there's some difference in how the minimizer uses (presumably the same) forces such that minimization fails for lots of systems?

There are some other alternatives that might work here. We can probably code up a simple gradient descent minimizer as a CustomIntegrator, but VariableLangevinIntegrator is pretty much an adaptive version of the same thing when the force is very large. Maybe give that a try?

**Comment by stefdoerr** [comment]
@jchodera How would I do that? I tried doing 1000 steps with the `VariableLangevinIntegrator` but it increased the potential energy like crazy.

```python
    platform, prop = getPlatform(platform, device)
    integrator = mm.VariableLangevinIntegrator(300 * unit.kelvin, 1.0 / unit.picoseconds, 0.001)#, 2.0 * unit.femtoseconds)
    simulation = app.Simulation(parm.topology, system, integrator, platform, prop)
    simulation.context.setPositions(coords)

    state = simulation.context.getState(getEnergy=True, getForces=True)
    print('Starting pot energy:', state.getPotentialEnergy().in_units_of(unit.kilocalories_per_mole))
    integrator.step(iterations)
    # simulation.minimizeEnergy(tolerance=minimizetol, maxIterations=iterations)
    state = simulation.context.getState(getPositions=True, getEnergy=True, getForces=True)
    print('Ending pot energy:', state.getPotentialEnergy().in_units_of(unit.kilocalories_per_mole))
```
```
Starting pot energy: 6072.946915383731 kcal/mol
Ending pot energy: 23992313895.866467 kcal/mol
```
Maybe I am just doing it wrong.

**Comment by peastman** [comment]
Set the temperature to 0.  You don't want to add heat to it.  I'd also use a larger friction coefficient.

**Comment by stefdoerr** [comment]
What is considered larger? I set temp to 0 K and coefficient 20 /
unit.picoseconds and it still increases

Starting pot energy: 6072.946936048214 kcal/mol
Ending pot energy: 9352.639695653632 kcal/mol
Starting pot energy: 6600.152449492716 kcal/mol
Ending pot energy: 10764.254062694685 kcal/mol
Starting pot energy: 5904.862478655719 kcal/mol
Ending pot energy: 11033.774060539006 kcal/mol

**Comment by peastman** [comment]
Try reducing the error tolerance.  You can also try increasing the friction coefficient further.  There's no universal "right" parameters.  If the energy is increasing at 0 K, that means it's not yet accurate enough for your purposes, so you need to tell it to be more accurate.

**Comment by peastman** [comment]
Fixed by #1984.
